api/src/utils/merge.js

```diff
@@ -1,92 +1,84 @@
+	await Promise.all([
+		...rhsFollows.map(f => f.removeFromStream()),
+		Follow.updateMany({ _id: { $in: update.map(f => f._id) } }, { rss: lhsID }),
+		Follow.remove({ _id: { $in: remove.map(f => f._id) } }),
+	]);
+	await Follow.getOrCreateMany(update.map(f => {
+		return { type: 'rss', userID: f.user._id, publicationID: lhsID };
+	}));
+}
+
+async function mergeArticlesAndPins(lhsID, rhsID) {
+	const rhsArticles = await Article.find({ rss: rhsID }).lean();
+	const rhsFingerprints = rhsArticles.map(a => a.fingerprint);
+	const lhsArticlesWithMatchingFingerprints = await Article.find({
+		rss: lhsID,
+		fingerprint: { $in: rhsFingerprints }
+	}).lean();
+	const lhsIDs = lhsArticlesWithMatchingFingerprints.map(a => a._id);
+	const lhsFingerprints = new Set(lhsArticlesWithMatchingFingerprints.map(a => a.fingerprint));
+
+	const update = rhsArticles.filter(a => !lhsFingerprints.has(a.fingerprint));
+	const remove = rhsArticles.filter(a => lhsFingerprints.has(a.fingerprint));
+
+	const [removePins, lhsPins] = await Promise.all([
+		Pin.find({ article: { $in: remove.map(a => a._id) } }),
+		Pin.find({ article: { $in: lhsIDs } })
+	]);
+
+	const lhsPinFingerprints = new Set(lhsPins.map(p => p.article.fingerprint));
+	const duplicatePins = removePins.filter(p => lhsPinFingerprints.has(p.article.fingerprint));
+	const newPins = removePins.filter(p => !lhsPinFingerprints.has(p.article.fingerprint));
+
+	const pinUpdates = newPins.map(p => {
+		const article = lhsArticlesWithMatchingFingerprints.filter(a => a.fingerprint == p.article.fingerprint)[0];
+		return Pin.updateOne({ _id: p._id }, { article });
+	});
+	await Promise.all([
+		Article.updateMany({ _id: { $in: update.map(f => f._id) } }, { rss: lhsID }),
+		...pinUpdates,
+		Article.remove({ _id: { $in: remove.map(f => f._id) } }),
+		Pin.remove({ _id: { $in: duplicatePins.map(f => f._id) } })
+	]);
+}
+
+async function mergeFeedUrls(lhsID, rhsID) {
+	const [lhs, rhs] = await Promise.all([
+		RSS.findById(lhsID).lean(),
+		RSS.findById(rhsID).lean()
+	]);
+	const feedUrls = new Set([lhs.feedUrl, rhs.feedUrl, ...lhs.feedUrls, ...rhs.feedUrls]);
+	await RSS.updateOne({ _id: lhsID }, { feedUrls: [...feedUrls] });
+}
 
-	// TODO: merge the feed url information
-	let feedUrls = [master.feedUrl].concat(
-		master.feedUrls,
-		[copy.feedUrl],
-		copy.feedUrls,
-	);
-	let uniqueUrls = {};
-	for (let url of feedUrls) {
-		uniqueUrls[url] = 1;
-	}
-	let newFeedUrls = Object.keys(uniqueUrls);
-	logger.info(`FeedUrls is now ${newFeedUrls}`);
-	master.feedUrls = newFeedUrls;
-	await master.save();
-	logger.info(
-		`Completed the merge. ${copy.feedUrl} is now merged with ${master.feedUrl}`,
-	);
+export async function mergeFeeds(lhsID, rhsID) {
+	lhsID = mongoose.Types.ObjectId(lhsID);
+	rhsID = mongoose.Types.ObjectId(rhsID);
 
-	return master;
+	await Promise.all([
+		mergeFollows(lhsID, rhsID),
+		mergeArticlesAndPins(lhsID, rhsID),
+		mergeFeedUrls(lhsID, rhsID)
+	]);
+    await RSS.remove({ _id: rhsID });
```

Comment:
Nope, this behaviour is wrong. We can't redirect on the frontend if we don't know what feed it was merged with @Nekuromento